BUG: interpolate.RegularGridInterpolator: fix linear interp. on 2D grid with a single point

miccoli opened the pull request:
#### Reference issue
Closes GH-23171

#### What does this implement/fix?
`RegularGridInterpolator` (method=linear) works also in the degenerate case of single-points grids.
However in the specific case of a 2D, float64 grid a Nan is returned.

Tests for generic single-point grids are added and the spurious case is fixed.

#### Additional information
Changes are kept to a bare minimum, however unnecessary branches (where `out[point] = NAN` is set) were simply discarded.

miccoli committed:
TST: interpolate.RegularGridInterpolator: test on single-point grid

New method `TestRegularGridInterpolator.test_length_one_axes` checks
that `interpolate.RegularGridInterpolator` works as expected for
methods 'linear' and 'nearest' if the grid contains a single point
(in all dimensions).

As of 4c1a4da81 this fails for the following parametrization
`TestRegularGridInterpolator::test_length_one_axes[linear-2-float64]`,
but is OK in all other cases.

See GH-23171.

miccoli committed:
BUG:interpolate.RegularGridInterpolator: fix linear interp. on 2D grid with a single point

Explicitly handle the case in which there is a single grid point on both
axes.

Branches in which `out[point] = NAN` is set are deleted because
redundant: out-of-bounds points/nans are handled by caller.

tylerjereddy commented on scipy/interpolate/tests/test_rgi.py:
```diff
@@ -335,6 +335,41 @@ def test_fillvalue_type(self):
+        fill = dtype(1/42)
+        nshape = (1, ) * ndim
+        values = np.array(val, dtype=dtype)
+        values.shape = nshape
```
perhaps a bit more compact/idiomatic to avoid manually setting the shape of the array?

<details>

```diff
--- a/scipy/interpolate/tests/test_rgi.py
+++ b/scipy/interpolate/tests/test_rgi.py
@@ -343,11 +343,9 @@ class TestRegularGridInterpolator:
         # for all methods parametrized above
         xi = np.array([0], dtype=dtype)
         points = (xi, ) * ndim
-        val = dtype(1/7)
-        fill = dtype(1/42)
-        nshape = (1, ) * ndim
-        values = np.array(val, dtype=dtype)
-        values.shape = nshape
+        val = np.asarray(1 / 7, dtype=dtype)
+        fill = np.asarray(1 / 42, dtype=dtype)
+        values = np.full((1,) * ndim, val, dtype=dtype)
```

</details>

miccoli replied:
Here setting the shape directly stresses the fact that, regardless of the number of dimensions `ndim`, the `values` array should have size=1: so it is first constructed as a 0D array which is promoted to `ndim`D by assigning the shape.

Similarly `val` and `fill` are meant to be numpy scalars and not 0D arrays... 

Currently numpy is a little bit confusing in the treatment of 0D arrays vs. scalars, so I prefer to be 100% explicit on what is meant.

ev-br replied:
numpy scalars vs 0D arrays are confusing indeed! That said, the way it's written now makes it more confusing then just creating the right-shaped array (have I mentioned that we're in the "matters of taste" territory already?)

> val and fill are meant to be numpy scalars and not 0D arrays

why though? You want to reject fill_value and values being numpy scalars? Then please add an explicit assertion that it raises. I wouldn't though, exactly because it's confusing in numpy, and it's not always easy to predict if something is a scalar or a 0D array. (and in the spirit of array API, consider actually just calling `np.asarray` on something that might be a scalar)

tylerjereddy commented on scipy/interpolate/tests/test_rgi.py:
```diff
@@ -335,6 +335,41 @@ def test_fillvalue_type(self):
+        sample = np.array([x0, x_nan, x_nan_0, x_oob], dtype=dtype)
+        wanted = np.array([val, np.nan, np.nan, fill], dtype=dtype)
+        result = interp(sample)
+        assert_array_equal(result, wanted)
```
It may not matter too much, but this test is a fair bit more complex looking than the one in your original ticket--is that just to make it more realistic than the original toy case?

miccoli replied:
These test are meant to expliciltly show the desired behaviour... I added them to be sure that my patch (expressly deleting `out[point] = NAN` clause) would not introduce other regressions.

ev-br replied:
Thank you for making sure of not introducing regressions, this is great! The ask to only split a single "assert four things in one line" into several assertions.

TBH I find it a bit surprising that these `out[point]=NAN` branches can be just removed. Will need to audit test coverage I suppose.

As an aside, this whole 2D fast path should be rewritten at some point, cf gh-22291. This is of course absolutely separate, am only mentioning it in case you'd be interested in working on it after this one :-)

miccoli replied:
> TBH I find it a bit surprising that these `out[point]=NAN` branches can be just removed. Will need to audit test coverage I suppose.

Nan and out-of-bounds are handled by the caller here:
https://github.com/scipy/scipy/blob/4c1a4da8110cc709dc2ff5050d882318fc649e20/scipy/interpolate/_rgi.py#L443-L448
so no need to add a sentinel in `out[point]`.

tylerjereddy reviewed (commented):
It seems like Evgeni was happy to receive a fix for this so I'll let him review as domain expert.

I added a few small comments/suggestions. I checked locally to confirm fail before/pass after the patch here for the new test, and the fact that you haven't had to modify any old/current tests should help for reviewer confidence.

ev-br reviewed (changes requested):
This looks like a good start. Still, 
- the cython implementation should neither `assert` nor raise. IIUC you want to check that if the grid is indeed a single point, then return either the `value` or a nan.
- the test is a little hard to read. IIUC, you are checking that the return values are correct for `xi` being a nan, `xi` being excatly the `point` and `xi` being away from it. If so, great! Please just rewrite the test to separate the cases.

miccoli commented on scipy/interpolate/_rgi_cython.pyx:
```diff
@@ -27,29 +27,29 @@ def evaluate_linear_2d(const double_or_complex[:, :] values, # cannot declare as
+    if grid[0].shape[0] == grid[1].shape[0] == 1:
+        # single grid point along both axes: no interpolation
+        for point in range(num_points):
+            assert indices[0, point] == indices[1, point] == -1
```
The above assert should never raise: it is added only to explicitly state an invariant that is assumed to be true by the current implementation.

I understand that such use of asserts is not common/idiomatic in python... but I felt authorized to add my assert looking at line 28, where there is indeed an assert of this very same nature:

https://github.com/scipy/scipy/blob/4c1a4da8110cc709dc2ff5050d882318fc649e20/scipy/interpolate/_rgi_cython.pyx#L19-L29

When learning C I was taught that asserts should be used only for debugging, and that it is good practice to leave them also in production code (which is typically compiled with -DNDEBUG).
The reasons:
- explicitly document the invariants which are assumed to be valid in some point of the code
- help debugging regressions if for some reason the invariants fail in some future release.

So actually, the above assert should never raise and could be simply deleted or commened out.
Please let me know if I should 
- delete this assert (but then also the one at line 28), 
- add a message that clarifies its debugging-only use
- comment it out.

miccoli commented on scipy/interpolate/tests/test_rgi.py:
```diff
@@ -335,6 +335,41 @@ def test_fillvalue_type(self):
+    @pytest.mark.parametrize("dtype", [np.float32, np.float64])
+    @pytest.mark.parametrize("ndim", [1, 2, 3])
+    @pytest.mark.parametrize("method", ["linear", "nearest"])
+    def test_length_one_axes(self, dtype, ndim, method):
```
Just a question of style: some projects implement a strict requirement of a single assert per unit test (this to have a single point of failure... instead of a sequnce of `assert A`, some code, `assert B`, some other code `assert C`).

Personally I like the idea of having a single assert, and indeed this was the reason for dumping all the tests (which are in my opinion relevant, to avoid future regressions).

Should I split the different cases as multiple asserts in the same method, or should I write multiple methods with a single assert?

miccoli replied:
OK, clear enough, hope to fix this soon.

miccoli committed:
Update scipy/interpolate/tests/test_rgi.py

Co-authored-by: Evgeni Burovski <evgeny.burovskiy@gmail.com>

miccoli committed:
MAINT: delete asserts from code

miccoli committed:
TST: refactor test method

- Rename `test_length_one_axes` as `test_length_one_axis_all`.
- Split tests for various types of input points into separate asserts.
- Use more idiomatic array API functions.

miccoli committed:
TST: delete spurious `assert_array_equal`

Make clear that it is the interpolator call that fails,
`not assert_array_equal`.
Use a slightly more verbose call pattern: best practice is to have
a bare assert, without inside calls.

miccoli committed:
Merge branch 'main' into feature/GH-23171

Conflicts:
- scipy/interpolate/tests/test_rgi.py
  Resolved by explicitly calling np.testing.assert_array_equal, and not
  using the (deleted) top module import

ev-br commented:
> Is there a preferred method for array equality testing

Please use `xp_assert_close` or `xp_assert_equal` for exact equality.

miccoli committed:
TST: Implement xp_assert_equal

- explicitly set wanted dtype
- cleanup minor code nits

ev-br reviewed (approved):
LGTM, thank you @miccoli. And congrats with the first scipy contribution, keep them coming!
